File: leetcode/backtracking_recursion/java_solutions/MinimumGeneticMutation_433.java
package leetcode.backtracking_recursion.java_solutions;

public class MinimumGeneticMutation_433 {

    static class Solution {
        private Integer min;

        public int minMutation(String start, String end, String[] bank) {
            min = null;
            backtrack(start, end, bank, new boolean[bank.length], 0);
            return min == null ? -1 : min;
        }

        private void backtrack(String current, String end, String[] bank, boolean[] visited, int path) {
            if (path > bank.length) {
                return;
            }
            if (current.equals(end)) {
                min = min == null ? path : Math.min(min, path);
                return;
            }
            for (int i = 0; i < bank.length; i++) {
                if (!visited[i] && isValidStep(current, bank[i])) {
                    visited[i] = true;
                    backtrack(bank[i], end, bank, visited, path + 1);
                    visited[i] = false;
                }
            }
        }

        private boolean isValidStep(String current, String bank) {
            if (current.length() != bank.length()) return false;
            int counter = 0;
            for (int i = 0; i < current.length(); i++) {
                if (current.charAt(i) != bank.charAt(i)) {
                    if (++counter > 1) {
                        return false;
                    }
                }
            }
            return true;
        }
    }
}

File: leetcode/backtracking_recursion/golang_solutions/minimum_genetic_mutation_433.go
package golang_solutions

var min int

func minMutation(start string, end string, bank []string) int {
	min = 1 << 30
	backtrack(start, end, bank, make([]bool, len(bank)), 0)
	return verifyResponse(min)
}

func backtrack(current string, end string, bank []string, visited []bool, path int) {
	if path > len(bank) {
		return
	}
	if current == end {
		min = minVal(path, min)
		return
	}
	for i := 0; i < len(bank); i++ {
		if !visited[i] && isValidStep(current, bank[i]) {
			visited[i] = true
			backtrack(bank[i], end, bank, visited, path+1)
			visited[i] = false
		}
	}
}

func isValidStep(start, bank string) bool {
	if len(start) != len(bank) {
		return false
	}
	counter := 0
	for i := range start {
		if start[i] != bank[i] {
			counter++
			if counter > 1 {
				return false
			}
		}
	}
	return true
}

func verifyResponse(val int) int {
	if val == 1<<30 {
		return -1
	}
	return val
}

func minVal(a, b int) int {
	if a > b {
		return b
	}
	return a
}
